src/client/js/client.js

```diff
@@ -1138,17 +1139,85 @@ define([
 
         function getModifiedNodes(newerNodes) {
             var modifiedNodes = [],
+                updatedMetaPaths = [],
+                metaNodes,
+                metaPath,
+                updatePath,
+                loadUnloadPath,
+                pathPieces,
                 i;
 
+            // For the client these rules apply for finding the affected nodes.
+            // 1. Updates should be triggered to any node that core.isTypeOf (i.e. mixins accounted for).
+            // 2. Root node should always be triggered.
+            // 3. loads/unloads should trigger updates for the parent chain.
+
+            if (state.loading.changedNodes) {
+                // 1. Account for mixins - i.e resolve isTypeOf.
+                // Gather all meta-nodes that had an update.
+                metaNodes = state.core.getAllMetaNodes(newerNodes[ROOT_PATH].node);
+                for (updatePath in state.loading.changedNodes.update) {
+                    if (metaNodes.hasOwnProperty(updatePath)) {
+                        updatedMetaPaths.push(updatePath);
+                    }
+                }
+
+                if (updatedMetaPaths.length > 0) {
+                    // There are meta-nodes with updates.
+                    for (metaPath in metaNodes) {
+                        // For all meta nodes..
+                        if (metaNodes.hasOwnProperty(metaPath)) {
+                            for (i = 0; i < updatedMetaPaths.length; i += 1) {
+                                // check if it is a typeOf (includes mixins) any of the updated meta-nodes
+                                if (state.core.isTypeOf(metaNodes[metaPath],
+                                        metaNodes[updatedMetaPaths[i]]) === true) {
+                                    // if so add its path to the update nodes.
+                                    state.loading.changedNodes.update[metaPath] = true;
+                                }
+                            }
+                        }
+                    }
+                }
+                //console.log('Update after meta considered', Object.keys(state.loading.changedNodes.update));
+
+                // 2. Add Root node
+                state.loading.changedNodes.update[ROOT_PATH] = true;
+
+                // 3. Account for loads and unloads.
+                for (loadUnloadPath in state.loading.changedNodes.load) {
+                    if (state.loading.changedNodes.load.hasOwnProperty(loadUnloadPath)) {
+                        pathPieces = loadUnloadPath.split('/');

```

Comment:
I think, as a general rule of thumb, we should avoid path manipulation outside core (if it is unavoidable then we should provide some functions by the core API rather than implementing manipulation in different places)
Also it seems that the loaded and unloaded nodes are coming from the diff, but the diff cannot compute all the new nodes (as well as all the removed ones).